Add 'always visible' channels to filtering mechanism

hacst opened the pull request:
Previously it was only possible to mark channels that are
supposed to get filtered when channel filtering is active.
This patch adds the capability to mark channels that ought
to never be filtered also. This is important if the "Hide
empty channels when filtering" feature is used.

To enable this we add an additional column to the channels_filtered
table which tracks our new possible channel filter configuration
states:
* Always visible when filtering:
  The channels is always visible even if it is empty
  and filtering empty channels is enabled.
* Never visible when filtering:
  The channel is never visible. It is filtered even if it
  contains users. The only exception is if the user itself
  is in a channel in which case it is not filtered.
* Normal/Default:
  The channel is hidden when filtering if it is empty and
  hiding empty channels when filtering is enabled.

The current version of this patch uses the filter_on.svg
and filter_off.svg icons as placeholders. Instead two new
icons ought to be chosen for this purpose.

This PR is meant as a basis for discussion and testing:
* Do we want this feature?
* Is the UI integration the best we can do (kinda hope not)?
* Is it robust?
* Does it work on all plattforms (only tested Linux)?
* Does upgrading work (ought to)?
* ???

Fixes  #1830

mkrautz commented:
I think this is fine, and I think the UI is OK -- it's the natural extension of the existing UI.

Then we can do extra UI work in another PR, if we want to make the feature easier to use.

Just throwing ideas out there:

 - The channel filter should have its own dialog.
 - Clicking the channel filter toolbar item should show this dialog, to make the feature more accessible.
 - The channel filter dialog should have a checkbox/switch near the top to toggle the channel filter.
 - The "hide empty channels" setting should be moved from UI settings to the channel filter's dialog.
 - The channel filter dialog should reproduce the channel tree and allow you to mark channels as "always show", "hide", or "none".
 - The "hide when filter is enabled" or the new context menu group should be removed in favour of using the dialog.

hacst commented:
Sounds good to me. The additional dialog is probably a good idea. On a big server with lots of channels it is quite hard to keep track of which channel you marked with what. Then again I hope most people are served by filtering empty channels and defining a few exceptions. The original idea was the simply use the filter toggle to join empty/filtered stuff but that might not be feasible/desirable on big servers.

Anyways. So I guess the concrete TODOs left for this PR are code-review, verifying osx and windows, adding two "new" icons to the default themes and updating the "what's this" (totally forgot about that).

For the icons we could even just use the same one for both and just consider it indication that there is "a" specific filter rule applied.

mkrautz commented on src/mumble/MainWindow.ui:
```diff
@@ -887,6 +890,39 @@ the channel's context menu.</string>
+    <string>&amp;Default</string>
+   </property>
+   <property name="toolTip">
+    <string>Apply channel default behavior.</string>
```
Maybe:

> Don't apply any special channel filtering rules to this channel. 

?

I dunno. "Apply default channel behavior" sounds confusing to me. I would expect it to explain what that default behavior is.

mkrautz commented on src/mumble/UserModel.cpp:
```diff
@@ -425,8 +429,10 @@ QVariant UserModel::data(const QModelIndex &idx, int role) const {
 
-				if (c->bFiltered)
-					l << (qiFilter);
+				if (c->filteredVisibility == Channel::FILTERED_VISIBILITY_NEVER)
```
`{}`?

mkrautz commented on src/mumble/UserModel.cpp:
```diff
@@ -425,8 +429,10 @@ QVariant UserModel::data(const QModelIndex &idx, int role) const {
-					l << (qiFilter);
+				if (c->filteredVisibility == Channel::FILTERED_VISIBILITY_NEVER)
+					l << (qiFilterVisibilityNever);
+				else if (c->filteredVisibility == Channel::FILTERED_VISIBILITY_ALWAYS)
```
`{}`?

mkrautz commented on src/mumble/UserModel.cpp:
```diff
@@ -1507,3 +1515,4 @@ void UserModel::updateOverlay() const {
 	g.o->updateOverlay();
 	g.lcd->updateUserView();
 }
+
```
Superfluous newline at the end? I see no newline complaint in the diff view here, so I assume there is now two newlines at the end?

mkrautz commented on src/mumble/UserView.cpp:
```diff
@@ -170,7 +170,7 @@ void UserView::mouseReleaseEvent(QMouseEvent *evt) {
 			hasComment = true;
 
-			if (channel->bFiltered)
+			if (channel->filteredVisibility != Channel::FILTERED_VISIBILITY_NORMAL)
```
`{}`?

mkrautz commented on src/mumble/UserView.cpp:
```diff
@@ -287,29 +287,33 @@ void UserView::selectSearchResult() {
-	{
-		if(channelHasUsers(c->qlChannels[i]))
-			return true;
+static bool isSessionUserInOrBelowChannel(const Channel* channel) {
```
`Channel *`

mkrautz commented on src/mumble/UserView.cpp:
```diff
@@ -287,29 +287,33 @@ void UserView::selectSearchResult() {
 	}
-	return false;
+	
+	const ClientUser* user = ClientUser::get(g.uiSession);
```
`ClientUser *`.

mkrautz commented on src/mumble/UserView.cpp:
```diff
@@ -287,29 +287,33 @@ void UserView::selectSearchResult() {
-			return true;
-		c=c->cParent;
+bool isChannelFiltered(const Channel *c) {
+	if(!g.s.bFilterActive) {
```
`if (`

mkrautz commented on src/mumble/UserView.h:
```diff
@@ -57,8 +57,8 @@ class UserView : public QTreeView {
-		
-	public slots:
+
+public slots:
```
(tab)public slots:

hacst commented:
Pushed a commit to address the review comments.

mkrautz commented on src/mumble/UserView.cpp:
```diff
@@ -287,35 +287,39 @@ void UserView::selectSearchResult() {
-	{
-		if(channelHasUsers(c->qlChannels[i]))
-			return true;
+static bool isSessionUserInOrBelowChannel(const Channel *channel) {
```
Doc comment?

mkrautz commented on src/mumble/UserView.cpp:
```diff
@@ -287,35 +287,39 @@ void UserView::selectSearchResult() {
-		if(c->bFiltered)
-			return true;
-		c=c->cParent;
+bool isChannelFiltered(const Channel *c) {
```
Doc comment? (should this be static?)

mkrautz commented:
@hacst Your turn. :-)

mkrautz commented:
@hacst Ping?

mkrautz commented:
Ping?

hacst commented:
Eh. Pong. Let's just pretend I didn't forget about this for a few months...

I addressed your two remaining comments, rebased and squashed. However I think the todo's I mentioned above remain:

- [ ] icons
- [ ] what's this
- [ ] testing on other platforms

Krzmbrzl commented:
Closing this as abandoned
